File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.py
def rotate(board, point):
    row1, col1 = point[0] - 1, point[1] - 1
    row2, col2 = point[2] - 1, point[3] - 1
    horizontal_len = abs(col2 - col1 + 1)
    vertical_len = abs(row2 - row1 + 1)
    perimeter = horizontal_len * 2 + vertical_len * 2 - 4

    r = row1
    c = col1 - 1
    arr = []
    for _ in range(horizontal_len):
        c += 1
        arr.append(board[r][c])
    for _ in range(vertical_len - 1):
        r += 1
        arr.append(board[r][c])
    for _ in range(horizontal_len - 1):
        c -= 1
        arr.append(board[r][c])
    for _ in range(vertical_len - 2):
        r -= 1
        arr.append(board[r][c])
    arr = [arr[-1]] + arr[0:-1]

    r = row1
    c = col1 - 1
    index = 0
    for _ in range(horizontal_len):
        c += 1
        board[r][c] = arr[index]
        index += 1
    for _ in range(vertical_len - 1):
        r += 1
        board[r][c] = arr[index]
        index += 1
    for _ in range(horizontal_len - 1):
        c -= 1
        board[r][c] = arr[index]
        index += 1
    for _ in range(vertical_len - 2):
        r -= 1
        board[r][c] = arr[index]
        index += 1
    return min(arr)


def solution(rows, columns, queries):
    answer = []
    board = [[0] * columns for _ in range(rows)]
    counter = 1
    for i in range(rows):
        for j in range(columns):
            board[i][j] = counter
            counter += 1
    for query in queries:
        answer.append(rotate(board, query))

    return answer

File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.java
class Solution {
	public int[] solution(int rows, int columns, int[][] queries) {
		int[] answer = new int[queries.length];

		int[][] board = new int[rows][columns];

		int cnt = 0;
		for (int r = 0; r < rows; r++) {
			for (int c = 0; c < columns; c++) {
				board[r][c] = ++cnt;
			}
		}

		for (int i = 0; i < queries.length; i++) {
			int[] query = queries[i];
			int r1 = query[0] - 1;// 2
			int c1 = query[1] - 1;// 2
			int r2 = query[2] - 1;// 5
			int c2 = query[3] - 1;// 4
			answer[i] = rotate(board, r1, c1, r2, c2);
		}

		return answer;
	}

	private int rotate(int[][] board, int r1, int c1, int r2, int c2) {
		int min = Integer.MAX_VALUE;

		int temp1 = board[r1][c2];
		for (int c = c2; c > c1; c--) {
			board[r1][c] = board[r1][c - 1];
			min = Math.min(min, board[r1][c]);
		}

		int temp2 = board[r2][c2];
		for (int r = r2; r > r1; r--) {
			if (r == r1 + 1) {
				board[r][c2] = temp1;
			} else {
				board[r][c2] = board[r - 1][c2];
			}
			min = Math.min(min, board[r][c2]);
		}

		int temp3 = board[r2][c1];
		for (int c = c1; c < c2; c++) {
			if (c == c2 - 1) {
				board[r2][c] = temp2;
			} else {
				board[r2][c] = board[r2][c + 1];
			}
			min = Math.min(min, board[r2][c]);
		}

		for (int r = r1; r < r2; r++) {
			if (r == r2 - 1) {
				board[r][c1] = temp3;
			} else {
				board[r][c1] = board[r + 1][c1];
			}
			min = Math.min(min, board[r][c1]);
		}

		return min;
	}
}